File: CPP/No213.cpp
/**
 * Created by Xiaozhong on 2020/8/21.
 * Copyright (c) 2020/8/21 Xiaozhong. All rights reserved.
 */
#include "iostream"
#include "vector"
#include "algorithm"

using namespace std;

class Solution {
private:
    int myRob(vector<int> &nums, int start, int end) {
        int pre = 0, cur = 0, tmp;
        for (int i = start; i < end; i++) {
            tmp = cur;
            cur = max(cur, pre + nums[i]);
            pre = tmp;
        }
        return cur;
    }

public:
    int rob(vector<int> &nums) {
        if (nums.empty()) return 0;
        if (nums.size() == 1) return nums[0];
        return max(myRob(nums, 0, nums.size() - 1),
                   myRob(nums, 1, nums.size()));
    }
};

int main() {
    vector<int> nums = {1, 2, 3, 1};
    Solution s;
    cout << s.rob(nums) << endl;
}

File: Java/src/No213.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

public class No213 {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        return Math.max(rob(nums, 0, n - 1), rob(nums, 1, n));
    }

    private int rob(int[] nums, int start, int end) {
        int[] dp = new int[nums.length + 1];
        dp[start] = 0;
        dp[start + 1] = nums[start];
        for (int i = start + 1; i < end; ++i) {
            dp[i + 1] = Math.max(dp[i], dp[i - 1] + nums[i]);
        }
        return dp[end];
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 1};
        No213 no213 = new No213();
        System.out.println(no213.rob(nums));
    }
}
